src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl

```diff
@@ -0,0 +1,195 @@
+using Oceananigans
+using Oceananigans.Grids
+using Oceananigans.Operators
+using Oceananigans.BuoyancyModels: buoyancy_perturbationᶜᶜᶜ
+using Oceananigans.Grids: AbstractGrid, AbstractUnderlyingGrid, halo_size
+using Oceananigans.ImmersedBoundaries
+using Oceananigans.Utils: getnamewrapper
+using Adapt 
+using Printf
+
+import Oceananigans.Grids: with_halo, znode
+import Oceananigans.Operators: Δzᶜᶜᶠ, Δzᶜᶜᶜ, Δzᶜᶠᶠ, Δzᶜᶠᶜ, Δzᶠᶜᶠ, Δzᶠᶜᶜ, Δzᶠᶠᶠ, Δzᶠᶠᶜ
+import Oceananigans.Advection: V_times_∂t_∂s_grid
+import Oceananigans.Architectures: arch_array
+
+"""
+    AbstractVerticalSpacing{R} 
+
+spacings for a generalized vertical coordinate system. The reference (non-moving) spacings are stored in `Δr`. 
+`Δ` contains the spacings associated with the moving coordinate system.
+`s⁻`, `sⁿ` and `∂t_∂s` fields are the vertical derivative of the vertical coordinate (∂Δ/∂Δr)
+at timestep `n-1` and `n` and it's time derivative.
+`denomination` contains the "type" of generalized vertical coordinate (the only one implemented is `ZStar`)
+"""
+abstract type AbstractVerticalSpacing{R} end
+
+const AbstractVerticalSpacingRG  = RectilinearGrid{<:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:AbstractVerticalSpacing} 
```

Comment:
why is the property all the way at the end of the list of properties of the grid? makes this harder to read